Refactor oncoprint module to include plotting by broad histology

cbethell opened the pull request:
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->

### Purpose/implementation Section

The purpose of this PR is to refactor the oncoprint module to include an option for plotting by broad histology per #968

#### What scientific question is your analysis addressing?

NA

#### What was your approach?

I included some logic in the `oncoprint-landscape` module bash script to iterate through the given broad histology values (currently supplied in an array in the bash script) and adapted to `01-plot-oncoprint.R` script to filter the metadata file, as well as the additional data files, to include only the data associated with the given broad histology.

I was able to create oncoprints for the categories specified on the associated issue https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/968#issuecomment-812175023
> It may look best if we create plots for the following 4 broad histologies and combine all of the others into Other CNS while keeping their broad_histology labels.
Low-grade astrocytic tumor
Embryonal tumor
Diffuse astrocytic and oligodendroglial tumor
Ependymal tumor

The broad histology goi lists should run in this script without a glitch once provided to the `goi_list` flag in the bash script.

Per another https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/968#issuecomment-812223409 left on the related issue, the `Normal` display group value is associated with `broad_histology` values of `NA` (a decision that was made when mapping the histology color labels https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/899#issuecomment-759504742), but no longer appears on the oncoprints as the "Other CNS" plots are now created by hardcoding the `broad_histology` categories we do want (the hard-coding is mostly due to some weird R behavior and in adhering to the preferred categories specified above and on the related issue).

#### What GitHub issue does your pull request address?

Part of #968 (this PR does not cover placing the plots in a panel form and the consensus maf file does not yet appear ready for implementation)

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?

The added broad histology plots will need a particularly close look.

#### Is there anything that you want to discuss further?

I do want to note that on the "Other CNS" oncoprints, the display group value `Low-grade astrocytic tumor` appears as it is associated with the `Other astrocytic tumor` broad histology. Is this intended behavior?
<img width="352" alt="Screen Shot 2021-04-02 at 10 17 34 PM" src="https://user-images.githubusercontent.com/43576623/113465453-0cc4c680-9402-11eb-826f-a947ca0690dd.png">

**Edit:** This behavior indeed results from the intentional decision made when mapping the histology color labels per https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/899#issuecomment-759634809.

> I think you can combine `Other astrocytic tumor` into LGAT for now, since it is a low-grade astrocytoma. Possibly, we can update that `broad_histology` in v19. 

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes.


### Results

#### What types of results are included (e.g., table, figure)?

Oncoprints for each broad histology.
The added plots include:
`all_participants_primary_only_diffuse-astrocytic-and-oligodendroglial-tumor_goi_oncoprint.png`
`all_participants_primary_only_diffuse-astrocytic-and-oligodendroglial-tumor_oncoprint.png`
`all_participants_primary_only_embryonal-tumor_goi_oncoprint.png`
`all_participants_primary_only_embryonal-tumor_oncoprint.png`
`all_participants_primary_only_ependymal-tumor_goi_oncoprint.png`
`all_participants_primary_only_ependymal-tumor_oncoprint.png`
`all_participants_primary_only_low-grade-astrocytic-tumor_goi_oncoprint.png`
`all_participants_primary_only_low-grade-astrocytic-tumor_oncoprint.png`
`all_participants_primary_only_other-cns_goi_oncoprint.png`
`all_participants_primary_only_other-cns_oncoprint.png`

`all_participants_primary-plus_diffuse-astrocytic-and-oligodendroglial-tumor_goi_oncoprint.png`
`all_participants_primary-plus_diffuse-astrocytic-and-oligodendroglial-tumor_oncoprint.png`
`all_participants_primary-plus_embryonal-tumor_goi_oncoprint.png`
`all_participants_primary-plus_embryonal-tumor_oncoprint.png`
`all_participants_primary-plus_ependymal-tumor_goi_oncoprint.png`
`all_participants_primary-plus_ependymal-tumor_oncoprint.png`
`all_participants_primary-plus_low-grade-astrocytic-tumor_goi_oncoprint.png`
`all_participants_primary-plus_low-grade-astrocytic-tumor_oncoprint.png`
`all_participants_primary-plus_other-cns_goi_oncoprint.png`
`all_participants_primary-plus_other-cns_oncoprint.png`


#### What is your summary of the results?

The plots appear to represent the accurate information and number of samples.

#### Reproducibility Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

<!-- Please review https://github.com/AlexsLemonade/OpenPBTA-analysis#documenting-your-analysis -->

- [ ] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

<!-- IF YOUR PULL REQUEST IS A DATA RELEASE, PLEASE REMOVE THE [HTML COMMENT TAG](https://html.com/tags/comment-tag/) FROM THE SECTION BELOW AND COMPLETE THE CHECKLIST-->

<!-- 
### Data Release Checklist
- [ ] Is the table in doc/data-file-descriptions.md up to date?
- [ ] Is doc/data-format.md up to date?
- [ ] Is doc/release-notes.md up to date?
- [ ] Is download-data.sh up to date?
- [ ] Was download-data.sh tested and did it complete without error?
-->

jaclyn-taroni commented on analyses/oncoprint-landscape/01-plot-oncoprint.R:
```diff
@@ -215,7 +260,7 @@ maf_object <- prepare_maf_object(
 # dataset and save as a png file.
 png(
-  file.path(plots_dir, opt$png_name),
+  file.path(plots_dir, tolower(gsub(" ", "-", opt$png_name))),
```
Good add 👍🏻 🚀

jaclyn-taroni commented on analyses/oncoprint-landscape/run-oncoprint.sh:
```diff
@@ -6,6 +6,9 @@
 set -e
 set -o pipefail
 
+# Print oncoprints by broad histology - will not by default
+HISTOLOGY=${RUN_BROAD_HISTOLOGY:-0}
```
I think we can take this logic for skipping splitting up by broad histology out. We know we probably want to break the plots up into separate panels. The way you've altered `01-plot-oncoprint.R` means one could just run without `--broad_histology` and get plots for the entire cohort if we wanted!

jaclyn-taroni commented on analyses/oncoprint-landscape/01-plot-oncoprint.R:
```diff
@@ -156,18 +162,13 @@ if (!is.null(opt$goi_list)) {
   goi_list <- unique(unlist(goi_list))
 }
 
-# Read in recurrent focal CNVs file
-if (!is.null(opt$focal_file)) {
-  focal_df <- readr::read_tsv(file.path(opt$focal_file))
-}
-
```
This is something that should have been removed earlier, correct?

cbethell replied:
Correct!

jaclyn-taroni commented on analyses/oncoprint-landscape/01-plot-oncoprint.R:
```diff
@@ -177,6 +178,50 @@ metadata <- metadata %>%
+        broad_histology %in% c(
+          "Tumors of sellar region",
+          "Neuronal and mixed neuronal-glial tumor",
+          "Tumor of cranial and paraspinal nerves	",
```
This extra spacing is required? 🤔

jaclyn-taroni commented on analyses/oncoprint-landscape/01-plot-oncoprint.R:
```diff
@@ -177,6 +178,50 @@ metadata <- metadata %>%
   dplyr::mutate(display_group = forcats::fct_reorder(display_group, display_order))
 
+# Filter to the metadata associated with the broad histology value, if provided
+if (!is.null(opt$broad_histology)) {
+  
+  if (!opt$broad_histology == "Other CNS") {
+    metadata <- metadata %>%
+      dplyr::filter(broad_histology == opt$broad_histology)
+  } else {
+    metadata <- metadata %>%
+      dplyr::filter(
+        broad_histology %in% c(
+          "Tumors of sellar region",
+          "Neuronal and mixed neuronal-glial tumor",
+          "Tumor of cranial and paraspinal nerves	",
+          "Meningioma",
+          "Mesenchymal non-meningothelial tumor	",
+          "Germ cell tumor",
+          "Choroid plexus tumor	",
+          "Histiocytic tumor",
+          "Tumor of pineal region",
+          "Metastatic tumors",
+          "Other astrocytic tumor",
+          "Lymphoma",
+          "Melanocytic tumor",
+          "Other tumor"
+        )
+      )
+  }
```
I think this logic works well for now, but am making a note that this is an area where we might want to take a different approach when we have different genes of interest lists. I don't think it makes sense to make any changes to accommodate #969 just yet!

jaclyn-taroni reviewed (commented):
This looks good! I had a few comments and questions, the most substantive of which is moving to separate plots by broad histology as our default.

cbethell commented on analyses/oncoprint-landscape/01-plot-oncoprint.R:
```diff
@@ -177,6 +178,50 @@ metadata <- metadata %>%
+  
+  # There's a duplicated `Tumor_Sample_Barcode` column that we will want to
+  # remove before filtering `maf_df`
+  maf_df <- maf_df[, -16]
```
> * Is this a new column or did we see this in other releases?

@jaclyn-taroni, this duplicate column was added to the `00-map-to-sample_id.R` script [here](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/6bdd22b504016e09df0d2801d79c34ddd6d83026/analyses/oncoprint-landscape/00-map-to-sample_id.R#L176) in merged PR #243 (https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/243#discussion_r345794392). 

My understanding is that the original `Tumor_Sample_Barcode` column in the maf file is needed to filter to tumor samples only as this column holds the `Kids_First_Biospecimen_ID` values https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/6bdd22b504016e09df0d2801d79c34ddd6d83026/analyses/oncoprint-landscape/00-map-to-sample_id.R#L127

However, because we use `sample_id` values in the `01-plot-oncoprint.R` script, the `Tumor_Sample_Barcode` column was recoded to hold the `sample_id` values
https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/6bdd22b504016e09df0d2801d79c34ddd6d83026/analyses/oncoprint-landscape/00-map-to-sample_id.R#L176

which resulted in a duplicated `Tumor_Sample_Barcode` column.

I initially tried to remove the original `Tumor_Sample_Barcode` column with the biospecimen IDs in the `00` script using:

`dplyr::select(-Tumor_Sample_Barcode) %>%` before the `rename` step linked above, but this initially produced an error for me, that I was interestingly unable to reproduce this morning so this may be a better solution here?

Note that it does not tell us it's duplicated, both columns are named `Tumor_Sample_Barcode`.

jaclyn-taroni replied:
> `dplyr::select(-Tumor_Sample_Barcode) %>%` before the `rename` step linked above, but this initially produced an error for me, that I was interestingly unable to reproduce this morning so this may be a better solution here?

My understanding is as follows: we want to replace the biospecimen IDs in the `Tumor_Sample_Barcode` column with `sample_id` so that we can match between the experimental strategies, _but_ we need to call the `sample_id` column `Tumor_Sample_Barcode` because that's a requirement of `oncoplot()`. That makes me think that that what you lay out that previously gave you an error is a good move!

cbethell commented:
In my [most recent commit](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/983/commits/27384147e6ce13aa3352d725bd0eceb72536bbb2) @jaclyn-taroni, I updating the logic handling the "duplicate" `Tumor_Sample_Barcode` column as I mentioned in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/983#discussion_r607980625, removed the flag added to the bash script as you suggested in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/983#discussion_r607333844 and removed that extra spacing that snuck into the hard-coded broad histology values.

Upon re-running, the "Other CNS" plots were updated due to that update in spacing (good catch!), otherwise the results stayed the same 👍.

All that said, I believe this is ready for another review!

jaclyn-taroni commented on analyses/oncoprint-landscape/run-oncoprint.sh:
```diff
@@ -72,18 +84,30 @@ Rscript --vanilla 00-map-to-sample_id.R \
-  --metadata_file ${histologies_file} \
-  --goi_list ${genes_list} \
-  --png_name ${primaryplus_filename}_goi_oncoprint.png
+  Rscript --vanilla 01-plot-oncoprint.R \
+    --maf_file ${intermediate_directory}/${primaryplus_filename}_maf.tsv \
+    --cnv_file ${intermediate_directory}/${primaryplus_filename}_cnv.tsv \
+    --fusion_file ${intermediate_directory}/${primaryplus_filename}_fusions.tsv \
+    --metadata_file ${histologies_file} \
+    --png_name ${primaryplus_filename}_"$histology"_oncoprint.png \
+    --broad_histology "$histology"
+
+  # Genes of interest only version of oncoprint
+  Rscript --vanilla 01-plot-oncoprint.R \
+    --maf_file ${intermediate_directory}/${primaryplus_filename}_maf.tsv \
+    --cnv_file ${intermediate_directory}/${primaryplus_filename}_cnv.tsv \
+    --fusion_file ${intermediate_directory}/${primaryplus_filename}_fusions.tsv \
+    --metadata_file ${histologies_file} \
+    --goi_list ${genes_list} \
+    --png_name ${primaryplus_filename}_"$histology"_goi_oncoprint.png \
+    --broad_histology "$histology"
```
Can we put this in the same loop as the primary only steps? I can't think of a reason not to at this stage.

cbethell replied:
Good point, done!

jaclyn-taroni reviewed (approved):
This looks good! One comment about putting the primary and the primary plus steps in the same loop, but no need for me to review!
